### PR TITLE
[NPM] Fix enforcing DNS/HTTP/Kafka stats limit

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

Fix state tracking code in the network_tracer module so that the `system_probe_config.max_dns_stats` config is enforced.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/state.go [hunk 1]
```diff
@@ -507,10 +507,10 @@ func (ns *networkState) storeDNSStats(stats dns.StatsByKeyByNameByType) {
 	// Fast-path for common case (one client registered)
 	if len(ns.clients) == 1 {
 		for _, c := range ns.clients {
-			if len(c.dnsStats) == 0 {
+			if len(c.dnsStats) == 0 && getDeepDNSStatsCount(stats) <= ns.maxDNSStats {
```
Comment (mbakht) [inline pkg/network/state.go:510]: is there any way to enforce that `stats` does not grow beyond the limit in the first place?

Comment (hmahmood) [reply pkg/network/state.go:510]: There is a separate limit elsewhere for that it looks like: https://github.com/datadog/datadog-agent/blob/hasan.mahmood/fix-dns-stats-limit/pkg/network/dns/snooper.go#L82

**Comment by guyarb** [comment]
Can we tell (and document in the PR description) what is the memory / cpu improvement from such a change?

**Review by guyarb — CHANGES_REQUESTED** [review]
If you are changing USM stats, then please run USM load test to ensure accuracy remains 100%

**Review by guyarb — APPROVED** [review]

File: pkg/network/state.go (hunk 1 above)
Comment (mbakht) [reply pkg/network/state.go:510]: I know this PR doesn't really add any new limit but merely fixes the enforcement of an existing one - but I am curious to know why we need both for DNS stats. If there is an obvious distinction that I am missing, addition of a comment may be helpful.

**Review by mbakht — APPROVED** [review]

File: pkg/network/state.go (hunk 1 above)
Comment (hmahmood) [reply pkg/network/state.go:510]: I think the limit in state.go is meant as an upper limit on how many stats we report in the connections check. If the other limit in snooper was > 75000 (the hardcoded state.go limit), we would drop data in state.go.

**Pull request merged**